hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +85,964 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
+                          MPI_LOR,
+                          m_exec_conf->getMPICommunicator());
+            }
+#endif // ENABLE_MPI
+
+        if (invalid_mass)
+            {
+            m_exec_conf->msg->warning() << "Some particles have a mass <= 0, may lead to "
+                                           "invalid results during MPCD collision."
+                                        << std::endl;
+            }
+        if (needs_thermostat && has_no_thermostat)
+            {
+            throw std::runtime_error("Thermostat required when colliding with rigid bodies.");
+            }
+        if (central_interacting)
+            {
+            m_exec_conf->msg->warning() << "Central particle of rigid body included in "
+                                           "MPCD collision. Check if this is intentional."
+                                        << std::endl;
+            }
+        if (invalid_center_of_mass)
+            {
+            throw std::runtime_error(
+                "Some rigid bodies do not have center of mass at central particle.");
+            }
+        if (invalid_mass_sum)
+            {
+            throw std::runtime_error(
+                "Mass of some rigid bodies not equal to sum of constituent particle masses.");
+            }
+        }
+    m_checked_collision_warnings = true;
+    }
+
+void mpcd::CollisionMethod::storeInitialEmbeddedGroupVelocities(uint64_t timestep)
+    {
+    const unsigned int num_group = m_embed_group->getNumMembers();
+    ArrayHandle<Scalar4> h_initial_vel(m_initial_velocity,
+                                       access_location::host,
+                                       access_mode::overwrite);
+    ArrayHandle<unsigned int> h_embed_group(m_embed_group->getIndexArray(),
+                                            access_location::host,
+                                            access_mode::read);
+    ArrayHandle<Scalar4> h_velocity(m_pdata->getVelocities(),
+                                    access_location::host,
+                                    access_mode::read);
+    for (unsigned int idx = 0; idx < num_group; ++idx)
+        {
+        // collect the initial velocities of the embedded particles
+        const unsigned int particle_index = h_embed_group.data[idx];
+        h_initial_vel.data[idx] = h_velocity.data[particle_index];
+        }
+    }
+
+void mpcd::CollisionMethod::thermalizeConstituentParticles(uint64_t timestep)
+    {
+    // zero accumulators
+    m_linmom_accum.zeroFill();
+    m_angmom_accum.zeroFill();
+
+    Scalar T_set = (*m_T)(timestep);
+
+    ArrayHandle<Scalar4> h_velocity(m_pdata->getVelocities(),
+                                    access_location::host,
+                                    access_mode::readwrite);
+    ArrayHandle<Scalar4> h_alt_vel(m_pdata->getAltVelocities(),
+                                   access_location::host,
+                                   access_mode::readwrite);
```

Comment:
These will be overwritten, right? (We don't care what value is in this array before this method.)
```suggestion
                                   access_mode::overwrite);
```

Reply:
I checked again and I noticed that we do read what is in this array in the last for loop. Does this mean it shouldn't be overwrite, or can you read values just find with the overwrite access mode as long as the value was overwritten at some point earlier in the method?

Reply:
Overwrite means the array should not be copied if it changes location (CPU to GPU or vice versa). After you access by overwrite, you only expect the values you write to be meaningful. The values of any other entries in the array are undefined.

In this case, you can’t expect the values of any entries in this array to be meaningful, so overwrite should be fine. Since the array stays in scope the whole time, it’s also OK to read these values later. If it went out of scope, you would need to reacquire it in one of the read modes.